rfcs/0000-rlp-v2.md

```diff
@@ -0,0 +1,1028 @@
+context.request...        → ...
+auth.<X>                  → metadata(key: "envoy.filters.http.ext_authz", path: <X>)
+ratelimit.domain          → <hostname>
+```
+
+...to yield effectively:
+
+```yaml
+rate_limits:
+  - actions:
+      - request_headers:
+          descriptor_key: "context.request.http.host"
+          header_name: ":authority"
+      - metadata:
+          descriptor_key: "auth.identity.username"
+          metadata_key:
+            key: "envoy.filters.http.ext_authz"
+            path:
+              - segment:
+                  key: "identity"
+              - segment:
+                  key: "username"
+```
+
+### Artificial Limitador condition for `matches`
+
+For each limit, the RLP controller will generate an artificial Limitador condition that ensures that the limit applies only when that one filterred matching rule is honoured to serve the request. This can be implemented with a 2-step procedure:
+1. generate an unique identifier for the limit – e.g. by applying any sufficiently entropic hash function of choice to the `matches` block of the limit;
+2. associate a `generic_key` type descriptor action to the each `HTTPRouteMatch` targeted by the limit.
+
+For example, given the following RLP ([Example 2](#example-2-specific-route-matching-rule-targeted-conditions-counter-qualifiers-and-multiple-rates)):
+
+```yaml
+apiVersion: kuadrant.io/v2beta1
+kind: RateLimitPolicy
+metadata:
+  name: toystore-non-admin-users
+spec:
+  targetRef:
+    group: gateway.networking.k8s.io
+    kind: HTTPRoute
+    name: toystore
+  limits:
+    - name: readers
+      matches:
+        - path:
+            type: PathPrefix
+            value: "/toys"
+          method: GET
+      rates:
+        - limit: 50
+          duration: 1
+          unit: minute
+      when:
+        - selector: auth.identity.group
+          operator: neq
+          value: admin
+
+    - name: writers
+      matches:
+        - path:
+            type: PathPrefix
+            value: "/toys"
+          method: POST
+      rates:
+        - limit: 5
+          duration: 1
+          unit: minute
+      when:
+        - selector: auth.identity.group
+          operator: neq
+          value: admin
+```
+
+Apart from the descriptor action (associated with both routes):
+
+```yaml
+- metadata:
+    descriptor_key: "auth.identity.group"
+    metadata_key:
+      key: "envoy.filters.http.ext_authz"
+      path:
+        - segment:
+            key: "identity"
+        - segment:
+            key: "group"
+```
+
+...and its corresponding Limitador condition:
+
+```
+auth.identity.group != "admin"
+```
+
+The following additional artificial descriptor actions will be generated:
+
+```yaml
+# associated with route GET /toys*
+- generic_key:
+    descriptor_key: "context.request.http.route_matcher"
+    descriptor_value: "2718701ec9bfd79132e58d92aed722489443094bf9c616e1b74361fe68360f05" # SHA256 hashing of [{"path":{"type":"PathPrefix","value":"/toys"},"method":"GET"}]
```

Comment:
So this is interesting, the whole `HTTPRouteRule` is semantically possibly more precise than "one the matches" within it… but I also see that it'd be less "stable", i.e. adding a `match` would change the limit's condition and as such its counters. With the (currently?) suffix added `--RLP` there can't be no conflicts to a `HTTPRouteRule` from two distinct `RateLimitPolicy`ies… Which is probably … good too? Two distinct RLPs targeting the same `HTTPRouteRule` would be distinct limits, with distinct counters and as such the most restrictive wins…

Reply:
I didn't know the conditions were part of the limit's identity.

> With the (currently?) suffix added `--RLP` there can't be no conflicts to a `HTTPRouteRule` from two distinct `RateLimitPolicy`ies… Which is probably … good too?

Correct. Two policies targeting the same matching rules (not necessarily same `HTTPRouteRule`) would still generate two distinct unique conditions. But that doesn't mean the RLP controller has to treat them as two different limits (two different counters). The identity of a limit is first defined by the RLP controller, then in Limitador. What the RLP controller spots as a "conflict" and how it acts on it (i.e. pushing two defs to Limitador or merging them into one first) is really up to the RLP controller and how it's programmed to capture the user intent.